src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs

```diff
@@ -459,72 +456,48 @@ private static ParameterDescriptor CreateParameterDescriptor(PropertyModel prope
             // without the constraint to match. For example, actions without an [Area] attribute on their
             // controller should not match when a value has been given for area when matching a url or
             // generating a link.
-            foreach (var constraintAttribute in action.RouteConstraints)
+            foreach (var kvp in action.RouteValues)
             {
-                if (constraintAttribute.BlockNonAttributedActions)
-                {
-                    removalConstraints.Add(constraintAttribute.RouteKey);
-                }
-
+                keys.Add(kvp.Key);
+                
                 // Skip duplicates
-                if (!HasConstraint(actionDescriptor.RouteConstraints, constraintAttribute.RouteKey))
+                if (!actionDescriptor.RouteValues.ContainsKey(kvp.Key))
                 {
-                    actionDescriptor.RouteConstraints.Add(new RouteDataActionConstraint(
-                        constraintAttribute.RouteKey,
-                        constraintAttribute.RouteValue));
+                    actionDescriptor.RouteValues.Add(kvp.Key, kvp.Value);
                 }
             }
 
-            foreach (var constraintAttribute in controller.RouteConstraints)
+            foreach (var kvp in controller.RouteValues)
             {
-                if (constraintAttribute.BlockNonAttributedActions)
-                {
-                    removalConstraints.Add(constraintAttribute.RouteKey);
-                }
+                keys.Add(kvp.Key);
 
                 // Skip duplicates - this also means that a value on the action will take precedence
-                if (!HasConstraint(actionDescriptor.RouteConstraints, constraintAttribute.RouteKey))
+                if (!actionDescriptor.RouteValues.ContainsKey(kvp.Key))
```

Comment:
Can you move this into `AddIfNotKeys(string key, string value)`? We repeat this several times in this type.